src/te_checks.c

```diff
@@ -140,6 +140,27 @@ struct check_result *check_unordered_perms(__attribute__((unused)) const struct
 
 }
 
+struct check_result *check_no_self(__attribute__((unused)) const struct check_data *data,
+                                   const struct policy_node *node)
+{
+	if (node->flavor != NODE_AV_RULE && node->flavor != NODE_XAV_RULE) {
+		return alloc_internal_error("Bad node type given to check C-007");
+	}
+	struct av_rule_data *av_data = node->data.av_data;
+
+	if (av_data->sources->next ||
+	    av_data->targets->next ||
+	    0 == strcmp(av_data->targets->string, "self") ||
+	    0 != strcmp(av_data->sources->string, av_data->targets->string) ||
+	    !look_up_in_decl_map(av_data->sources->string, DECL_TYPE) ) {
```

Comment:
Maybe also issue on attributes? `allow domain self:dir search_dir_perms;`

```
!look_up_in_decl_map(av_data->sources->string, DECL_TYPE)
```

->

```
(!look_up_in_decl_map(av_data->sources->string, DECL_TYPE) && !look_up_in_decl_map(av_data->sources->string, DECL_ATTRIBUTE))
```

Reply:
More specifically: What is the reason to exclude all non known identifiers, e.g. `$1` or `$1_t` ?

Reply:
The problem with attributes is that "allow attr attr" and "allow attr self" are not equivalent. If foo_t and bar_t both associate with attr, then "allow attr self" expands to "allow foo_t foo_t" and "allow bar_t bar_t", but it does not include "allow foo_t bar_t" or "allow bar_t foo_t". "allow attr attr" on the other hand does include the latter two.

The refpolicy variable case I hadn't thought of, but it would have the problem that we don't know if they are attributes or not.

Since the attribute mistake has actual consequences, it would be nice to detect it, but I'm not sure of a way to infer probable develop intent off the top of my head.

Reply:
$1_t would be reasonable. I can add that case, although I wonder how common such a case would be.

Reply:
I've added the $*_t.  Can you check and make sure I got the boolean logic right?